generalizing rules

wurDevTim opened the pull request:
A request for two small changes in the rule engine. 
The reason for these is that at WUR we would like to run python based rules on the server. 

1. Support for different types of rules the 'irods_rule_engine_plugin-irods_rule_language-instance' is really cool. However, I hope many user, like me, will also like the 'irods_rule_engine_plugin-python-instance'. The new implementation defaults to the 'irods_rule_language', but it gives user the option to work with the other plugins.
2. Currently the rule file must be on the users machine. It would also be nice to have the ability to run rules which are available on the server. With this small change users can do both. The naming is taken from the python irods client, not saying it's the clearest.

Curious to hear what you think.

chStaiger commented:
>     1. Support for different types of rules the 'irods_rule_engine_plugin-irods_rule_language-instance' is really cool. However, I hope many user, like me, will also like the 'irods_rule_engine_plugin-python-instance'. The new implementation defaults to the 'irods_rule_language', but it gives user the option to work with the other plugins.

I have never tried the python-plugin for rules. We can allow for it, but this means that our integration tests also need to check that. Hence the irods docker images need to be updated and the tests in the docker need to be extended.

chStaiger commented:
>     2. Currently the rule file must be on the users machine. It would also be nice to have the ability to run rules which are available on the server. With this small change users can do both. The naming is taken from the python irods client, not saying it's the clearest.
Reading the documentation of the python-irodsrulenegine it does not look like you can ececute a rule file located on the server. 
The rule.py allows to use all rules on the server `core.re` and other configured rule engines. But the user still has to embed them in an own local file.

luijs commented:
The python rule language engine blocks usage of pythonrules for non rodsadmin users. We ported the rules that we make available to general users to core.re. 

Option 1 is still interesting for users that want to write their own rules and execute them against the irods python plugin. However, those users need to be rodsadmin to be allowed to do this.

For option 1, the rule execution with icommands is as follows:
``` irule -r irods_rule_engine_plugin-irods_rule_language-instance rule_name_in_core_re '*var1=foobar' ruleExecOut``` 


(To be complete, below the implementation in the core:)
core.re:
```
rule_name_in_core_re(){
   py_rule_name_in_core_re(*var1);
}
``` 
core.py:
```
def py_rule_in_core_re(rule_args, callback, rei):
    var1: str = rule_args[0]
    logic
```

wurDevTim commented:
> Reading the documentation of the python-irodsclient it does not look like you can execute a rule file located on the server. The rule.py allows to use all rules on the server `core.re` and other configured rule engines. But the user still has to embed them in an own local file.

Indeed, to run '.py' rules we have on the server we wrap them in '.r' rules and call these like:
`std_out, std_err = self.ic.execute_rule(body = 'irods_rdm_npec_create_experiment', params = params, rule_type = 'irods_rule_language') `

At the same time, it is an easy change to enable users to send `irods_rule_engine_plugin-python-instance` files to the server. 
If I get the option between writing the rule in python rule or the 'irods' rule language, I would always use the python implementation.

qubixes commented:
I am not familiar with the iRods rule system (at least not to any degree of expertise), but I imagine there are multiple/many ways to interact with the system. In that case I would suggest to use `**kwargs` to pass these on to the python-irodsclient. The named keyword arguments would be the most common use case, and the rest goes in the `**kwargs`. I think in that case we can skip adding those extra use cases to our test suite. Would that work?

wurDevTim commented:
Works for me, I don't think you have to test all possible combinations... But it would be nice if iBridges does not limit the users to sending rule files in the irods rule language.  

The key change would be checking if keywords like `rule_type` are already in the `**kwargs` and only setting them if needed.
@chStaiger  do you agree? Than I can rewrite the PR.

chStaiger commented:
> Works for me, I don't think you have to test all possible combinations... But it would be nice if iBridges does not limit the users to sending rule files in the irods rule language.
> 
> The key change would be checking if keywords like `rule_type` are already in the `**kwargs` and only setting them if needed. @chStaiger do you agree? Than I can rewrite the PR.

I would prefer if we extend the rule.py to also work with the python rule engine to check that in the docker. I.e. we have our current integration test with the iRODS rule language and we would need a second one for the python rule. Otherwise we cannot guarantee that the new feature really works.

chStaiger commented:
@wurDevTim if you want to finish this PR without the integration tests then please make sure to link in the tutorials to the relevant documentation in the PRC so that users are not completely left alone when things should fail ;)

wurDevTim commented:
Rewrote the code to use `kwargs`, updated the quickstart and tested a few rules.
For as far as I can see everything works.

As for documentation I now wrote the common keywords in the functions description to give people some hints.
And added an example in the quickstart to give an idea of how iBirdges could be used to run rules on the server.
I hope this suffices for now?

chStaiger commented:
I cannot merge before all of our tests pass ... You need to take look at that.

wurDevTim commented:
> I cannot merge before all of our tests pass ... You need to take look at that.
Checked with mypy, it looks like this shifted the failure to an issue in the integration pipeline itself:
![docker error](https://github.com/UtrechtUniversity/iBridges/assets/81219429/8e917181-528a-49b1-b899-8ba4b023d063)

qubixes commented on ibridges/rules.py:
```diff
@@ -25,21 +24,31 @@ def execute_rule(session: Session, rule_file: str, params: dict,
     tuple
         (stdout, stderr)
 
     """
```
It would also be nice if you have some examples to add them to the docstring.

wurDevTim replied:
For as far as I can see you are using the quickstart and other tutorials for this?

qubixes replied:
Yes, you're right that currently we're using mostly the quickstart/tutorials for that. I think in this case rules is not something that first time users will have to worry about, which is why I suggested putting it in the docstring/docs on readthedocs. It's also okay the way it is now, since our docs (and especially the API part) need some love at some point and that can be done at that point. I'll just have to remember to check this PR.

qubixes reviewed (changes requested):
Hi @wurDevTim, thanks for the PR, I think overall it is nice to add some features to ibridges, especially when it doesn't add a lot of maintenance cost on our side. I have some suggestions for the function signature and to make the API on readthedocs read a little bit more nicely.

wurDevTim commented on ibridges/rules.py:
```diff
@@ -25,21 +24,31 @@ def execute_rule(session: Session, rule_file: str, params: dict,
-        Rule arguments.
-    output : str
-        Rule output variable(s).
+    kwargs : dict example keywords:
```
updated

wurDevTim commented:
@qubixes  thanks for the feedback, it is insightful. I update the code. To make it work like you propose, the integration tests will also need an update. 

From:
```
    def test_rules(session, testdata):
        rule_fp = str(testdata/"example.r")
       execute_rule(session, rule_fp, {})
```

To:
```
    def test_rules(session, testdata):
        rule_fp = str(testdata/"example.r")
       execute_rule(session, rule_file=rule_fp, params={})
```

qubixes commented:
> @qubixes thanks for the feedback, it is insightful. I update the code. To make it work like you propose, the integration tests will also need an update.
> 
> From:
> 
> ```
>     def test_rules(session, testdata):
>         rule_fp = str(testdata/"example.r")
>        execute_rule(session, rule_fp, {})
> ```
> 
> To:
> 
> ```
>     def test_rules(session, testdata):
>         rule_fp = str(testdata/"example.r")
>        execute_rule(session, rule_file=rule_fp, params={})
> ```

Yes, the integration test would need to be updated. That is why I suggested to keep the old arguments/keyword arguments as well, see my first comment. Is it okay if you change it into that suggestion, then the integration test should start working as well.

wurDevTim commented:
@qubixes updated to your prefered solution. This PR makes me feel like a rookie, but after someone restarted the tests they all passed.

Within iRODS both the `params ` and the `rule_file` are optional, it's up to you if you want to leave them as optional keywords or add them to the kwargs later.

qubixes commented:
@wurDevTim  I restarted the integration tests, because they are not very stable. Looks good to me now, thanks for the effort to get all the tests passing, etc. I hope I wasn't too overbearing! I'll merge your PR now.
